src/trace/trace.c

```diff
@@ -96,18 +98,120 @@ static inline void mtrace_event(const char *data, uint32_t length)
 }
 
+#if CONFIG_TRACE_FILTERING_VERBOSITY
 /**
- * \brief Runtime trace filtering
+ * \brief Runtime trace filtering based on verbosity level
  * \param lvl log level (LOG_LEVEL_ ERROR, INFO, DEBUG ...)
  * \param uuid uuid address
  * \return false when trace is filtered out, otherwise true
  */
-static inline bool trace_filter_pass(uint32_t lvl,
-				     const struct tr_ctx *ctx)
+static inline bool trace_filter_verbosity(uint32_t lvl, const struct tr_ctx *ctx)
 {
 	/* LOG_LEVEL_CRITICAL has low value, LOG_LEVEL_VERBOSE high */
 	return lvl <= ctx->level;
 }
+#endif /* CONFIG_TRACE_FILTERING_VERBOSITY */
+
+#if CONFIG_TRACE_FILTERING_ADAPTIVE
+struct recent_log_entry {
+	uint32_t entry_id;
+	uint64_t message_ts;
+	uint64_t first_suppression_ts;
+	uint32_t suppress_count;
+};
+
+static struct recent_log_entry recent_entries[CONFIG_TRACE_RECENT_ENTRIES_COUNT];
+static struct recent_log_entry entry_to_emit;
+
+static inline void prepare_to_emit(struct recent_log_entry *entry)
+{
+	if (entry->suppress_count)
+		memcpy_s(&entry_to_emit, sizeof(entry_to_emit), entry, sizeof(*entry));
+	memset(entry, 0, sizeof(*entry));
+}
+
+static inline void emit_recent_entry(void)
+{
+	if (!entry_to_emit.suppress_count)
+		return;
+
+	_log_message(trace_log_unfiltered, false, LOG_LEVEL_INFO, _TRACE_INV_CLASS, &dt_tr,
+		     _TRACE_INV_ID, _TRACE_INV_ID, "Suppressed %u similar messages: %s",
+		     entry_to_emit.suppress_count, entry_to_emit.entry_id);
+
+	memset(&entry_to_emit, 0, sizeof(entry_to_emit));
+}
+
+static inline void emit_recent_entries(uint64_t current_ts)
+{
+	int i;
+
+	if (entry_to_emit.entry_id)
+		emit_recent_entry();
+
+	for (i = 0; i < CONFIG_TRACE_RECENT_ENTRIES_COUNT; i++) {
+		if (current_ts - recent_entries[i].message_ts >
+			CONFIG_TRACE_RECENT_TIME_THRESHOLD) {
+			prepare_to_emit(&recent_entries[i]);
+			emit_recent_entry();
+		}
+	}
+}
+
+/**
+ * \brief Runtime trace flood suppresion
+ * \param log_level log verbosity level
+ * \param entry ID of log entry
+ * \param message_ts message timestamp
+ * \return false when trace is filtered out, true otherwise
+ */
+static inline bool trace_filter_flood(uint32_t log_level, uint32_t entry, uint64_t message_ts)
+{
+	struct recent_log_entry *oldest_entry = recent_entries;
+	int i;
+
+	/* don't attempt to suppress debug messages using this method, it would be uneffective */
+	if (log_level >= LOG_LEVEL_DEBUG)
+		return true;
+
+	/* check if same log entry was sent recently */
+	for (i = 0; i < CONFIG_TRACE_RECENT_ENTRIES_COUNT; i++) {
+		/* in the meantime identify the oldest entry, for reuse later */
+		if (oldest_entry->first_suppression_ts > recent_entries[i].first_suppression_ts)
+			oldest_entry = &recent_entries[i];
+
+		if (recent_entries[i].entry_id == entry) {
+			if (message_ts - recent_entries[i].first_suppression_ts >=
+			    CONFIG_TRACE_RECENT_MAX_TIME) {
+				/* entry was suppressed long enough, time to release it */
+				prepare_to_emit(&recent_entries[i]);
```

Comment:
Instead of your `prepare_to_emit()` / `emit_recent_entry()` pair of functions I think you just want one that would do:
```
static evict_entry(struct recent_log_entry *entry)
{
	struct recent_log_entry tmp = *entry;

	memset(entry, 0, sizeof(*entry));
	if (tmp.suppress_count)
		_log_message(trace_log_unfiltered, false, LOG_LEVEL_INFO, _TRACE_INV_CLASS, &dt_tr,
				_TRACE_INV_ID, _TRACE_INV_ID, "Suppressed %u similar messages: %ToBeClarified",
				tmp.suppress_count, tmp.entry_id);
}
```
and then you don't need `static struct recent_log_entry entry_to_emit;` either

Reply:
@akloniex what about this one?

Reply:
@lyakh @lgirdwood The reason why I separated these functions was to maintain chronological order, to have current trace message, that's being sent, arrive before information about entries that are unsuppressed due to filter state update, caused by that message. This requires that extra step, to delay emitting entry from filter update until next message arrives, and then emit all entries that needed to be emitted up to this point.